docs/async/async-csharp-vb.rst

```diff
@@ -0,0 +1,200 @@
+		
+		Dim json As String = JsonConvert.SerializeObject(dirs);
+		
+		Return json
+	End Function
+	
+More on Task and Task<T>
+------------------------
+
+As mentioned before, Tasks are constructs which represent long-running operations calling code must wait on.
+
+* ``Task`` represents a single operation which does not return a value.
+* ``Task<T>`` represents a single operation which returns a value of type ``T``.
+
+It's important to reason about Tasks as abstractions of work to be done rather than threads, largely because they typically aren't threads at all!  Asynchronous .NET library calls essentially delegate work to the operating system, which is naturally asynchronous when I/O is performed.  Let this be made clear once more: when calling an async .NET library method, there is no new thread spawned to handle the work.
+
+Tasks are awaitable, meaning that the application of the ``await`` keyword will allow for either:
+
+* "Unwrapping" the return value for a ``Task<T>`` operation after it completes.
+* Simply waiting for a ``Task`` operation to finish before moving forward.
+
+Tasks are also used outside of the async programming model.  They are the foundation of the Task Parallel Library, which allows support for `Data Parallelism <https://msdn.microsoft.com/en-us/library/dd537608(v=vs.110).aspx>`_ and `Task Parallelism <https://msdn.microsoft.com/en-us/library/dd537609(v=vs.110).aspx>`_.  However, from the perspective of writing responsive code with the C#/VB async programming model, ``Task`` and ``Task<T>`` are really just used as a means to coordinating blocking operations and extracting their return values (if they have them).
+
+Important Info and Advice
+-------------------------
+
+Although async programming is relatively straightfoward, there are some details to keep in mind which could otherwise result in some nasty behavior.
+
+* **Best practice is to add "Async" to the end of every async method name you write which could be consumed by another method.**
+
+Failure to do so could result in having to track down a race condition later.  It's better to be explicit here!  Note that certain methods which aren't explicity called by your code (such as event handlers or web controller methods) may not necessarily apply.
+
+* ``await`` **is what will ultimately make a method asynchronous.**
+
+Failing to apply the ``await`` operator to a task will result in the async method running synchronously!  Application of ``await`` is what suspends the async method, giving back control to the method which called it.  Pay attention to compiler warnings about this.
+
+* ``async void`` **should only be used for event handlers.**
+
+Why?  That's the only reason they were allowed in the first place!  Async programming uses the ``Task`` and ``Task<T>`` objects, which provide flexibility in coordinating asynchronous work.  Throwing that out of the window with ``async void`` doesn't follow the model very well.  Here's some specific issues:
```

Comment:
You might clarify that this pattern is required to make the scenario work.

I don't think of the types as "providing flexibility", but representing async work in a first class way.

Try to avoid language like "throwing that out of the windows". It's not clear what you mean.